File: c++/KokoEatingBananas.cpp
// This code solves LeetCode question "875. Koko Eating Bananas", level: Medium
// The approach uses Binary Search
// The complexity analysis is---> TC: O(NlogN), SC: O(1)

#include<bits/stdc++.h>
using namespace std;

long long timeToEatBananas(int k, int n, vector<int>piles)
{
    long long sum=0;
    for(int i=0; i<n; i++)
    {
        if(piles[i]%k!=0)
            sum+=((piles[i]/k)+1);
        else
            sum+=(piles[i]/k);
    }
    return sum;
}
int minEatingSpeed(vector<int>& piles, int h) {
    int n=piles.size();
    int low=1, high=*max_element(piles.begin(), piles.end());
    int k=high;
    while(low<=high)
    {
        int mid=(low+high)/2;
        long long h1=timeToEatBananas(mid, n, piles);
        if(h1<=(long long)h)
        {
            k=min(k,mid);
            high=mid-1;
        }
        else if(h1>h)
            low=mid+1;
    }
    return k;        
}


int main()
{
    int n,h; 
    vector<int> piles;
    cout<<"Enter the number of piles: "<<endl;
    cin>>n;
    cout<<"Enter the total number of hours Koko has to eat all the bananas:"<<endl;
    cin>>h;
    cout<<"Enter the number of bananas in each pile:"<<endl;
    for(int i=0; i<n; i++)
    {
        int x; cin>>x;
        piles.emplace_back(x);
    }
    int k=minEatingSpeed(piles,h);
    cout<<"The bananas-per-hour eating speed to finish all bananas is: "<<k;
    
}

File: java/Koko_Eating_Bananas.java
// Below is the Java implementation for the problem "Koko Eating Bananas" using binary search to find the minimum eating speed.
// Problem link: https://leetcode.com/problems/koko-eating-bananas/description/
// Time Complexity: O(n log m) where n is the number of piles and m is the maximum number of bananas in a pile.
// Space Complexity: O(1)
// This code solves the problem efficiently by narrowing down the possible eating speeds using binary search.

class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int max = 0;
        int min = Integer.MAX_VALUE;

        // finding the maximum pile to set the upper limit for binary search
        for (int b : piles) {
            if (b > max) {
                max = b;
            }
        }

        // setting the search space for binary search
        long l = 1, r = max;

        // implementing binary search to find the minimum eating speed
        while (l <= r) {
            long mid = (l + r) / 2;

            long currhr = verify(mid, piles);

            if (currhr > h) {
                l = mid + 1;
            } else {
                // updating the minimum speed found so far so that we have the least possible speed
                min = (int) Math.min(min, mid);
                r = mid-1;
            }
        }

        return min;
    }

    // helper function to calculate the hours needed at a given speed
    private long verify(long num, int[] piles) {
        long hr = 0;

        for (int p : piles) {
            hr += (p + num - 1) / num;
        }

        return hr;
    }
}